temporal-sdk/src/main/java/io/temporal/internal/statemachines/ActivityStateMachine.java

```diff
@@ -288,20 +289,57 @@ private void setStartedCommandEventId() {
     startedCommandEventId = currentEvent.getEventId();
   }
 
-  private void cancelCommandNotifyCanceled() {
+  // *Immediately versions don't wait for a matching command, underlying callback will trigger an
+  // event loop so the workflow can make progress, because promise gets filled.
+
+  /**
+   * {@link CommandType#COMMAND_TYPE_SCHEDULE_ACTIVITY_TASK} command is not yet left to the server.
+   * Cancel it in place and immediately notify the workflow code.
+   */
+  private void cancelCommandNotifyCanceledImmediately() {
     cancelCommand();
     if (cancellationType != ActivityCancellationType.ABANDON) {
-      notifyCanceled();
+      notifyCanceled(false);
+    }
+  }
+
+  /**
+   * Workflow code doesn't need to wait for the cancellation event if {@link
+   * ActivityCancellationType#TRY_CANCEL}, immediately notify the workflow code.
+   */
+  private void notifyCanceledIfTryCancelImmediately() {
+    if (cancellationType == ActivityCancellationType.TRY_CANCEL) {
+      notifyCanceled(false);
```

Comment:
The abandon handling looks a little odd to me here and elsewhere.

Abandoned activities should still have the workflow code get notified that they were cancelled - or is the idea that this has already happened at a higher level?

If so, probably makes more sense to just deal with not calling the callback inside `notifyCanceled` itself

Reply:
Yes, Abandon is handled even earlier, on the entry point of requesting cancelation: https://github.com/temporalio/sdk-java/blob/7f2679c4ac53b19f6dfe54913cc13ad0dec84179/temporal-sdk/src/main/java/io/temporal/internal/statemachines/ActivityStateMachine.java#L280

Reply:
But it's actually an argument for simplifying the code and not even checking for `ABANDON` here.

Reply:
I add this TODO here

```
  private void cancelCommandNotifyCanceledImmediately() {
    cancelCommand();
    // TODO With {@link ActivityCancellationType#ABANDON} we shouldn't even get here as it gets
    //  handled in #cancel.
    //  It's a code path for TRY_CANCEL and WAIT_CANCELLATION_COMPLETED only.
    //  Was the original design to cancel a not-yet-sent
    //  {@link CommandType#COMMAND_TYPE_SCHEDULE_ACTIVITY_TASK} in case of ABANDON too?
    if (cancellationType != ActivityCancellationType.ABANDON) {
      notifyCanceled(false);
    }
  }
```

and not gonna touch it in this PR. It looks like this if for ABANDON is not needed, we are not getting here with ABANDON. But I feel like by cleaning it out I may lose some original context here.